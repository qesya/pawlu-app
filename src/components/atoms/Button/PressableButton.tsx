import React, { memo } from "react";
import {
  GestureResponderEvent,
  Pressable,
  StyleProp,
  ViewStyle,
} from "react-native";
import Animated, {
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from "react-native-reanimated";

type PressableButtonProps = {
  onPress: () => void;
  onPressIn?: (event: GestureResponderEvent) => void;
  onPressOut?: (event: GestureResponderEvent) => void;
  style?: StyleProp<ViewStyle>;
  children?: React.ReactNode;
  testID?: string;
  disabled?: boolean;
  pressableStyle?: StyleProp<ViewStyle>;
};

const PressableButton: React.FC<PressableButtonProps> = ({
  onPress,
  children,
  style,
  testID,
  disabled,
  onPressIn,
  onPressOut,
  ...props
}) => {
  const scale = useSharedValue(1);

  const animatedStyle = useAnimatedStyle(() => ({
    transform: [
      { scale: withSpring(scale.value, { damping: 10, stiffness: 200 }) },
    ],
  }));

  const handlePressIn = (e: GestureResponderEvent) => {
    scale.value = 0.95;
    onPressIn?.(e);
  };

  const handlePressOut = (e: GestureResponderEvent) => {
    scale.value = 1;
    onPressOut?.(e);
  };

  const buttonStyles: ViewStyle = {
    flexDirection: "column",
    alignContent: "center",
  };

  return (
    <Pressable
      onPress={onPress}
      onPressIn={handlePressIn}
      onPressOut={handlePressOut}
      testID={testID}
      disabled={disabled}
      style={props.pressableStyle}
      {...props}
    >
      <Animated.View
        style={[buttonStyles, style, animatedStyle]}
        testID="pressable-btn-inner-view-test"
      >
        {children}
      </Animated.View>
    </Pressable>
  );
};

export default memo(PressableButton);
