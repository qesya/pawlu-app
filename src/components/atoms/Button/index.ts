export { default } from "./Button";
export { default as PressableButton } from "./PressableButton";
