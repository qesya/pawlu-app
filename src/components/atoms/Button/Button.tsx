import { Icon } from "@/src/components";
import { useColors } from "@/src/hooks";
import { BUTTON_SIZES, ButtonSize, ButtonVariants } from "@/src/theme";
import React, { memo, useState } from "react";
import {
  ActivityIndicator,
  StyleProp,
  StyleSheet,
  TextStyle,
  View,
  ViewStyle,
} from "react-native";
import {
  useAnimatedStyle,
  useSharedValue,
  withSpring,
} from "react-native-reanimated";
import { BUTTON_ICON_TYPES } from "../Icon/Icon";
import Typography from "../Typography";
import PressableButton from "./PressableButton";

type ButtonProps = {
  type: ButtonVariants;
  title: string;
  buttonSize?: ButtonSize;
  onPress: () => void;
  style?: StyleProp<ViewStyle>;
  isLoading?: boolean;
  isDisabled?: boolean;
  icon?: keyof typeof BUTTON_ICON_TYPES;
  iconPosition?: "left" | "right";
  iconWidth?: number;
  iconHeight?: number;
  customIconColor?: string;
  customIconColorPressed?: string;
  textStyle?: StyleProp<TextStyle>;
  testId?: string;
  customGapIcon?: number;
};

type ButtonThemeProps = {
  defaultColor: string;
  hoverColor: string;
  disabledColor: string;
  textColorDefault: string;
  textColorDisabled: string;
  hoverTextColor?: string;
  borderColor?: string;
};

const Button: React.FC<Readonly<ButtonProps>> = ({
  type,
  title,
  onPress,
  buttonSize = "md",
  style,
  isLoading = false,
  isDisabled = false,
  icon,
  iconPosition = "left",
  iconWidth,
  iconHeight,
  customIconColor,
  customIconColorPressed,
  textStyle,
  testId,
  customGapIcon,
}) => {
  const {
    primaryColor400,
    primaryColor500,
    gray50,
    gray100,
    gray200,
    gray300,
    gray700,
    error500,
    error600,
    error200,
    error100,
    white,
  } = useColors();

  const scale = useSharedValue(1);
  const [isPressed, setIsPressed] = useState(false);
  const gapIconSize = customGapIcon ?? 8;

  const animatedStyle = useAnimatedStyle(() => ({
    transform: [
      { scale: withSpring(scale.value, { damping: 10, stiffness: 200 }) },
    ],
  }));

  const handlePressIn = () => {
    setIsPressed(true);
    scale.value = 0.95;
  };

  const handlePressOut = () => {
    setIsPressed(false);
    scale.value = 1;
  };

  const buttonTheme: Record<ButtonVariants, ButtonThemeProps> = {
    primary: {
      defaultColor: primaryColor400,
      hoverColor: primaryColor500,
      disabledColor: gray300,
      textColorDefault: white,
      textColorDisabled: white,
    },
    secondary: {
      defaultColor: gray100,
      hoverColor: gray200,
      disabledColor: gray200,
      textColorDefault: gray700,
      textColorDisabled: white,
    },
    tertiary: {
      defaultColor: "transparent",
      hoverColor: "transparent",
      disabledColor: "transparent",
      textColorDefault: primaryColor400,
      textColorDisabled: gray300,
      hoverTextColor: primaryColor500,
    },
    "destructive-primary": {
      defaultColor: error500,
      hoverColor: error600,
      disabledColor: error200,
      textColorDefault: white,
      textColorDisabled: white,
    },
    "destructive-secondary": {
      defaultColor: error100,
      hoverColor: error200,
      disabledColor: error200,
      textColorDefault: error500,
      textColorDisabled: white,
    },
    outline: {
      defaultColor: white,
      hoverColor: gray50,
      disabledColor: gray200,
      textColorDefault: gray700,
      textColorDisabled: gray300,
      borderColor: gray200,
    },
  };

  const fontSizeMap = {
    sm: "textSM",
    md: "textSM",
    lg: "textBase",
    xl: "textBase",
  } as const;

  const buttonHeight = BUTTON_SIZES[buttonSize];
  const fontSize = fontSizeMap[buttonSize];
  const fontWeight = "semiBold";

  const theme = buttonTheme[type];

  let backgroundColor = theme.defaultColor;

  if (isDisabled) {
    backgroundColor = theme.disabledColor;
  } else if (isPressed) {
    backgroundColor = theme.hoverColor;
  }

  let resolvedTextColor = theme.textColorDefault;

  if (isDisabled) {
    resolvedTextColor = theme.textColorDisabled ?? theme.textColorDefault;
  } else if (isPressed && "hoverTextColor" in theme) {
    resolvedTextColor = (theme as any).hoverTextColor;
  }

  const shouldApplyShadow = [
    "primary",
    "secondary",
    "destructive-primary",
    "destructive-secondary",
  ].includes(type);

  const buttonStyles: ViewStyle = {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
    backgroundColor,
    paddingHorizontal: 20,
    borderRadius: 8,
    height: buttonHeight,
    ...(type === "outline" && {
      borderWidth: 1,
      borderColor: theme.borderColor,
    }),
    ...(shouldApplyShadow && {
      shadowColor: "rgba(18, 23, 33, 1)",
      shadowOpacity: 0.05,
      shadowOffset: { width: 0, height: 1 },
      shadowRadius: 2,
      elevation: 1,
    }),
  };

  return (
    <PressableButton
      onPress={onPress}
      onPressIn={handlePressIn}
      onPressOut={handlePressOut}
      disabled={isDisabled}
      style={[buttonStyles, style, animatedStyle]}
      testID={testId}
    >
      {isLoading ? (
        <ActivityIndicator
          testID="activity-indicator"
          size="small"
          color={theme.textColorDefault}
        />
      ) : (
        <View style={styles.btnOuter}>
          {icon && iconPosition === "left" && (
            <Icon
              icon={icon}
              height={iconHeight}
              width={iconWidth}
              fill={
                isPressed
                  ? (customIconColorPressed ?? resolvedTextColor)
                  : (customIconColor ?? resolvedTextColor)
              }
              style={{ marginRight: gapIconSize }}
            />
          )}
          <Typography
            style={[
              {
                color: resolvedTextColor,
                textDecorationLine: type === "tertiary" ? "underline" : "none",
              },
              textStyle,
            ]}
            weight={fontWeight}
            size={fontSize}
          >
            {title}
          </Typography>
          {icon && iconPosition === "right" && (
            <Icon
              icon={icon}
              height={iconHeight}
              width={iconWidth}
              fill={
                isPressed
                  ? (customIconColorPressed ?? resolvedTextColor)
                  : (customIconColor ?? resolvedTextColor)
              }
              style={{ marginLeft: gapIconSize }}
            />
          )}
        </View>
      )}
    </PressableButton>
  );
};

const styles = StyleSheet.create({
  btnOuter: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
  },
});

export default memo(Button);
