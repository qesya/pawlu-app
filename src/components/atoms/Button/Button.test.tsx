import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import Button from "../Button";

jest.mock("@/src/hooks", () => ({
  useColors: () => ({
    primaryColor400: "#1a73e8",
    primaryColor500: "#1669c1",
    gray50: "#f9f9f9",
    gray100: "#f1f1f1",
    gray200: "#e2e2e2",
    gray300: "#d1d1d1",
    gray700: "#333333",
    error500: "#e53935",
    error600: "#d32f2f",
    error200: "#ef9a9a",
    error100: "#ffcdd2",
    white: "#ffffff",
  }),
  useResponsiveFontSize: (size: number) => size,
}));

jest.mock("react-native-reanimated", () =>
  require("react-native-reanimated/mock"),
);

describe("Button", () => {
  it("renders button title", () => {
    const { getByText } = render(
      <Button title="Click Me" type="primary" onPress={() => {}} />,
    );
    expect(getByText("Click Me")).toBeTruthy();
  });

  it("triggers onPress when pressed", () => {
    const onPressMock = jest.fn();
    const { getByTestId } = render(
      <Button
        title="Submit"
        type="primary"
        onPress={onPressMock}
        testId="button"
      />,
    );
    fireEvent.press(getByTestId("button"));
    expect(onPressMock).toHaveBeenCalledTimes(1);
  });

  it("does not trigger onPress when disabled", () => {
    const onPressMock = jest.fn();
    const { getByTestId } = render(
      <Button
        title="Disabled"
        type="primary"
        isDisabled
        onPress={onPressMock}
        testId="disabled-button"
      />,
    );
    fireEvent.press(getByTestId("disabled-button"));
    expect(onPressMock).not.toHaveBeenCalled();
  });

  it("shows loader when isLoading is true", () => {
    const { getByTestId, queryByText } = render(
      <Button
        title="Load"
        type="primary"
        isLoading
        onPress={() => {}}
        testId="loading-button"
      />,
    );
    expect(getByTestId("activity-indicator")).toBeTruthy();
    expect(queryByText("Load")).toBeNull();
  });

  it("renders left icon when provided", () => {
    const { toJSON } = render(
      <Button
        title="With Icon"
        type="primary"
        icon="arrow-right-icon"
        iconPosition="left"
        onPress={() => {}}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders right icon when provided", () => {
    const { toJSON } = render(
      <Button
        title="With Icon"
        type="primary"
        icon="arrow-right-icon"
        iconPosition="right"
        onPress={() => {}}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
