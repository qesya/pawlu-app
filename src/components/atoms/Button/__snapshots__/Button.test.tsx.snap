// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders left icon when provided 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      [
        {
          "alignContent": "center",
          "flexDirection": "column",
        },
        [
          {
            "alignItems": "center",
            "backgroundColor": "#1a73e8",
            "borderRadius": 8,
            "elevation": 1,
            "flexDirection": "row",
            "height": 40,
            "justifyContent": "center",
            "paddingHorizontal": 20,
            "shadowColor": "rgba(18, 23, 33, 1)",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.05,
            "shadowRadius": 2,
          },
          undefined,
          {
            "transform": [
              {
                "scale": 1,
              },
            ],
          },
        ],
        {
          "transform": [
            {
              "scale": 1,
            },
          ],
        },
      ]
    }
    testID="pressable-btn-inner-view-test"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={24}
        bbWidth={24}
        fill="none"
        focusable={false}
        height={24}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "marginRight": 8,
            },
            {
              "flex": 0,
              "height": 24,
              "width": 24,
            },
          ]
        }
        vbHeight={24}
        vbWidth={24}
        width={24}
      >
        <RNSVGGroup
          fill={null}
          propList={
            [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M4 12h16m0 0-6-6m6 6-6 6"
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
                "strokeLinecap",
                "strokeLinejoin",
              ]
            }
            stroke={
              {
                "payload": 4294967295,
                "type": 0,
              }
            }
            strokeLinecap={1}
            strokeLinejoin={1}
            strokeWidth={2}
          />
        </RNSVGGroup>
      </RNSVGSvgView>
      <Text
        style={
          [
            {
              "fontFamily": "InterSemibold",
              "fontSize": 14,
              "lineHeight": 17.5,
            },
            {
              "textAlign": "left",
            },
            undefined,
            [
              {
                "color": "#ffffff",
                "textDecorationLine": "none",
              },
              undefined,
            ],
          ]
        }
      >
        With Icon
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button renders right icon when provided 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      [
        {
          "alignContent": "center",
          "flexDirection": "column",
        },
        [
          {
            "alignItems": "center",
            "backgroundColor": "#1a73e8",
            "borderRadius": 8,
            "elevation": 1,
            "flexDirection": "row",
            "height": 40,
            "justifyContent": "center",
            "paddingHorizontal": 20,
            "shadowColor": "rgba(18, 23, 33, 1)",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.05,
            "shadowRadius": 2,
          },
          undefined,
          {
            "transform": [
              {
                "scale": 1,
              },
            ],
          },
        ],
        {
          "transform": [
            {
              "scale": 1,
            },
          ],
        },
      ]
    }
    testID="pressable-btn-inner-view-test"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "InterSemibold",
              "fontSize": 14,
              "lineHeight": 17.5,
            },
            {
              "textAlign": "left",
            },
            undefined,
            [
              {
                "color": "#ffffff",
                "textDecorationLine": "none",
              },
              undefined,
            ],
          ]
        }
      >
        With Icon
      </Text>
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={24}
        bbWidth={24}
        fill="none"
        focusable={false}
        height={24}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "marginLeft": 8,
            },
            {
              "flex": 0,
              "height": 24,
              "width": 24,
            },
          ]
        }
        vbHeight={24}
        vbWidth={24}
        width={24}
      >
        <RNSVGGroup
          fill={null}
          propList={
            [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M4 12h16m0 0-6-6m6 6-6 6"
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            propList={
              [
                "stroke",
                "strokeWidth",
                "strokeLinecap",
                "strokeLinejoin",
              ]
            }
            stroke={
              {
                "payload": 4294967295,
                "type": 0,
              }
            }
            strokeLinecap={1}
            strokeLinejoin={1}
            strokeWidth={2}
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
</View>
`;
